pkg/ccl/changefeedccl/changefeed_processors.go

```diff
@@ -683,17 +665,44 @@ func (ca *changeAggregator) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMet
 	}
 
 	if !ca.shutdownCheckpointEmitted {
-		// Aggregator shutdown may be initiated by the coordinator.
-		// Emit an up-to-date frontier information in case coordinator will restart.
-		ca.shutdownCheckpointEmitted = true
-		ca.AppendTrailingMeta(execinfrapb.ProducerMetadata{
-			Changefeed: &execinfrapb.ChangefeedMeta{Checkpoint: getFrontierSpans()},
-		})
+		// Best effort attempt to emit shutdown checkpoint.
+		// Even if this call returns an error, we cannot do anything
+		// with this error (such as ca.MoveToDraining(err)) since the processor
+		// is already in the "StateTrailingMetadata" (and thus panics if you attempt
+		// to MoveToDraining again).
+		_ = ca.emitShutdownCheckpoint(false /* not a node drain */)
 	}
 
 	return nil, ca.DrainHelper()
 }
 
+func (ca *changeAggregator) emitShutdownCheckpoint(isDrain bool) error {
+	// Before emitting shutdown checkpoint, we must flush any buffered events.
+	if err := ca.flushBufferedEvents(); err != nil {
```

Comment:
It would be good to know how fast this happens realistically. The node drain mechanism waits for 10 seconds before cancelling the context. A concern here is that flushing events may take more than 10 seconds, in which case we would get an error here and return it.

It is still better to have this code than not to have it though.

Reply:
It's all best effort; if this is not fast enough -- one can change the flags to increase the timeout.